### PR TITLE
bounds_transformer could bypass global_bounds due to the test logic within _trim function in domain_reduction.py

### PR DESCRIPTION
**Pull request by perezed00**
Previously, when the upper limit of  new_bounds < lower limit of global_bounds, the new_bounds could bypass the global_bounds.  The proposed changes requires the little change to the existing code, but could cause one of the entries in new_bounds to stagnate at the global_bounds if the new_bounds entry is always outside of the global_bounds.

### CODE REVIEW

**Comment by till-m** [comment]
This should've been fixed in #398 (c.f. also #394). I don't think this change has been deployed to PyPI yet though.

**Comment by perezed00** [comment]
Thanks for linking me to those posts, I was not aware of them. However, this issue is distinct from  #394, which addresses a window-width bypass. In `1.4.3` the global_bounds can still be bypassed when the suggested upper bound is smaller than the original lower bound. This PR addresses the current issue by identifying if either of the suggested new_bounds are outside of the original_bounds, in which case it sets the suggested new_bound value to the global_bound value. 

This issue arises, for instance, when an optimizer instance loads a log that was generated by an instance with bounds greater than those of the current instance. 

I will return to this PR to add a test case to `tests/test_seq_domain_red.py` and to verify the order of the various checks in `_trim`. I suspect that the proposed solution would be more robust if the entries of new_bounds are sorted before the global bounds are enforced.

**Comment by till-m** [comment]
Hey @perezed00,

just to confirm, the case you're trying to cover is that sometimes, the lower bound can be greater than the upper bound (i.e. the bounds are out of order) and then the lower bound is not trimmed properly and allowed to be greater than the global bound?

**Comment by perezed00** [comment]
No, the case I've identified is roughly as follows:
Suppose the global bounds for a variable are [5, 10]. If `bounds_transformer` suggests new_bounds = [1, 4], then _both_ bounds are smaller than the lower global_bound. `v1.4.3` only checks if _one_ of the bounds is smaller than global limit, so the second bound sneaks in.

Suppose the global_bounds for a variable are [x_lower, x_upper]. 
Suppose bounds_transformer suggests new_bounds=[nx_lower, nx_upper]. Assume nx_lower<nx_upper.

If nx_lower < x_lower, `v1.4.3` catches the error.
If (nx_lower< x_lower  and  nx_upper < x_lower), `v1.4.3` does not catch the error. (Caught by this PR)
If (nx_lower > nx_upper  and nx_upper < x_lower), needs to be checked.

**Comment by till-m** [comment]
Thanks, that makes sense! I guess some more testing is in order. If the case you describe only shows up when loading logs from a differently configured optimizer, it makes sense that it wasn't covered before.
I assume you now have a solid understanding of the trim logic. Would you mind adding some comments to the code as part of this PR to make it clear what's happening in the `_trim` function?

**Comment by perezed00** [comment]
Sounds good. I don't mind adding some comments to the `_trim` function. 
The case I've described can also arise after the use of `optimizer.set_bounds(new_bounds={})` 

In summary, I'll revise this PR after doing the following:
- Add comments to `_trim`
- Verify the (nx_lower > nx_upper and nx_upper < x_lower) case described above
- Check if the entries of new_bounds should be sorted before global bound enforcement
- Add a test case to `tests/test_seq_domain_red.py`

It will probably be a few days before I get a chance to implement these, but I'll post again when complete.

**Comment by till-m** [comment]
Hey @perezed00, did you get around to have another look at this? Seems like the issue might've popped up for another user so I'd be eager to get this fixed.

**Comment by perezed00** [comment]
Hi, @till-m 

I have updated this pull request. Regarding the list of tasks above, here's what was done:
- Comments were added to `_trim`. The section of `_trim` that deals with the window size could still use more comments, however.
- The (nx_lower > nx_upper and nx_upper < x_lower) case described above is tested, and not and issue
- The entries of new bounds are now sorted before global bounds are enforced. This change was not **needed** but it makes the organization of `_trim` a little easier to follow
- Test cases were added to `tests/test_seq_domain_red.py`

Additionally, the name of variables in `_trim` was made more uniform. With the exception of `test_notebooks_run.py` all of the tests in `/tests` pass. `test_notebooks_run.py` was not tested.

File: bayes_opt/domain_reduction.py
Comment (till-m) [inline bayes_opt/domain_reduction.py:105]: I think this can malfunction since iterating over an array (`__next__`) returns a pointer to the original object's data.

```python
import numpy as np
new_bounds = np.array([[2., 1.], [4., 3.]])
for i, pbounds in enumerate(new_bounds):
    if pbounds[0] > pbounds[1]:
        new_bounds[i, 0] = pbounds[1]
        new_bounds[i, 1] = pbounds[0]
```
Out:
```
array([[1., 1.],
       [3., 3.]])
```

File: bayes_opt/domain_reduction.py
Comment (till-m) [inline bayes_opt/domain_reduction.py:114]: This can essentially cause a reset to the initial global bounds if we pan outside them completely.
1. Can this case occur naturally, or only if the user manually messes with the bounds?
2. If it can happen naturally, can the bounds transformer recover from this and resume normal operation?

**Review by till-m — COMMENTED** [review]
Looks good to me in general. Some small questions and things to verify.

File: bayes_opt/domain_reduction.py
Comment (perezed00) [reply bayes_opt/domain_reduction.py:105]: I had a suspicion this could be a problem, but that's how it was previously implemented, so I left it as-is. Anyways, this can be handled by np.sort(pbounds) instead. I'll make the change.

File: bayes_opt/domain_reduction.py
Comment (perezed00) [reply bayes_opt/domain_reduction.py:114]: FYI, I think this possible route to non-convergence might already be present in `v1.4.3`

File: bayes_opt/domain_reduction.py
Comment (till-m) [reply bayes_opt/domain_reduction.py:105]: Yeah, you're right, it must've slipped by the last time. Thanks for handling it!

**Comment by till-m** [comment]
Regarding your alternative solution -- I like your suggestion, maybe the appropriate way to implement this would be to restrict `TargetSpace.max()` to return the maximum obeying the global bounds. The problem is, that the global bounds are currently only stored in the transformer, not in the target space. (cc: @fmfn @bwheelz36 @leandrobbraga)

Also, I think it'd be good to have a warning to inform the user that the transformer observed a maximum outside the global bounds and why this can happen.

File: bayes_opt/domain_reduction.py
Comment (perezed00) [reply bayes_opt/domain_reduction.py]: done

**Comment by perezed00** [comment]
Since we have identified a few separate issues, I propose we implement the solutions in separate PRs. 

Here is how I see the issues:
`Issue #0`: The transformer is suggesting boundaries that exceed the global bounds. 
`Issue #1`: Some erroneous boundaries do not get caught by existing `_trim` logic.
`Issue #2`: Current solutions to issue `#1` can prevent the optimizer from converging in certain cases. 

`Issue #0 `is the root cause of issues `#1` and `#2`.  This PR addresses `#1`, but leaves `#0` and `#2` unresolved. The [suggested ](https://github.com/bayesian-optimization/BayesianOptimization/pull/441#issuecomment-1707115417)solution above may address `#0` directly. For the sake of organized updates lets keep this discussion focused on `#1` and make a new PR for `#0`.  What do you think? If you agree I can make a new branch / PR.

I don't mind if this PR get's closed without integration (I'll leave that decision to you), I just don't want this single PR to become a discussion that is too long to follow.

**Comment by till-m** [comment]
Hi @perezed00,

my apologies, for not getting back to you, I have been rather busy recently.
I think it is generally better to tackle issues at the root. How would your solution to issue #0 look like (mathematically speaking)? I think there are several ways of doing what you're trying to accomplish with subtle, but important differences.

**Comment by perezed00** [comment]
Hi @till-m,

No problem, I have also been very busy. Basically, the if the transformer sees a set of parameters in the target_space that optimize the function, then it tries to converge to that set of parameters. Mathematically, can mask the target_space so that the transformer only sees the parameters within the global bounds.  I implemented a windowing function within `domain_reduction.py' that doesn't require modifying other portions of the code base. I'll upload that now, along with some improved comments. 

During this process, I learned that the mask will not stop the transformer from suggesting bounds outside of the `global_bounds`, since that process also depends on the zoom/pan parameters. I am not sure that it can be avoided, and I don't think it needs to be addressed. This is because the `_trim` function enforces the global bounds.

On the other hand, as far as I can tell, the mask does drastically improve convergence. I believe that a transformer with a mask is as good as (or better than) a "normal" transformer at converging.  This can (should?) be explicitly tested.

If, in your opinion, the convergence seems okay, we might want to suppress the convergence warning since it is quite common at this point.

File: bayes_opt/domain_reduction.py
Comment (till-m) [inline bayes_opt/domain_reduction.py:103]: I think it would be better to ensure that `target_space.max()` respects the pbounds and then use that function to get the maximum. Otherwise we will have two functions doing pretty much the same thing, which adds maintenance overhead (and the SDRT should really respect constraints, too).

File: bayes_opt/domain_reduction.py
Comment (till-m) [inline bayes_opt/domain_reduction.py:156]: `optimizer` generally refers to the bayesian optimizer, which should still converge to something, just slower compared to when the SDRT is used. This message might be confusing for users who could think that the entire optimization procedure has broken down. You could write, e.g.:

```suggestion
                    The offensive boundary has been reset, but the optimizer may proceed without domain reduction.""")
```
but not super attached to it and there's probably a better way to phrase that.

File: bayes_opt/domain_reduction.py
Comment (till-m) [inline bayes_opt/domain_reduction.py:163]: as above. Note that both of them state `lower bound has exceeded its global limit` but one of the cases presumably concerns the upper bound.

File: bayes_opt/domain_reduction.py
Comment (till-m) [inline bayes_opt/domain_reduction.py:190]: I quite like this code and the variable names are lovely :)

**Review by till-m — CHANGES_REQUESTED** [review]
Sorry for the delay...
Biggest thing from my side is the fact that the `max_window` should be calculated using the target space. Rest looks good to me :)

**Comment by perezed00** [comment]
Sorry for the late reply here , I am in the final stretch of dissertation work. I will finish this in 5 days.

**Comment by till-m** [comment]
No worries. Good luck with the dissertation!

**Comment by till-m** [comment]
Hey @perezed00, don't mean to rush you, but any update on this? :)

File: bayes_opt/domain_reduction.py
Comment (perezed00) [reply bayes_opt/domain_reduction.py:103]: This has been implemented. Thanks for the suggestion. I like it better

File: bayes_opt/domain_reduction.py
Comment (perezed00) [reply bayes_opt/domain_reduction.py:156]: I rewrote when the warnings were triggered to be more specific and reduce excessive warnings. Please see updated version.

**Comment by perezed00** [comment]
Hey @till-m ,

Thanks for your patience. The changes have been implemented. target_space.max() now respects the bounds and I improved the warning messages. Warnings are now more specific and less frequent. Let me know if you see any other issues.

Cheers

File: tests/test_seq_domain_red.py
Comment (perezed00) [inline tests/test_seq_domain_red.py:190]: this is consistent with the other test files.

File: tests/test_target_space.py
Comment (perezed00) [inline tests/test_target_space.py:205]: added local pbounds since the `.target_max()` method would now throw an error.

**Comment by till-m** [comment]
hey, I'm a bit swamped right now, will have a look as soon as I can. Thanks for the contributions :)

File: tests/test_seq_domain_red.py
Comment (till-m) [reply tests/test_seq_domain_red.py:190]: I think these should be removed, as they are not needed anymore, but it's good that you made sure it's consistent. Let's keep it like this for now, and we can delete them all in the future.

File: tests/test_target_space.py
Comment (till-m) [reply tests/test_target_space.py:170]: yeah, I think this works as intended. I saw that the case where there is a better point outside of pbounds is also covered, so I think this looks good 👍

File: tests/test_target_space.py
Comment (till-m) [inline tests/test_target_space.py:196]: nitpick: indentation of empty line

File: bayes_opt/target_space.py
Comment (till-m) [inline bayes_opt/target_space.py]: Is this function unused? If no, what's its purpose?

File: tests/test_target_space.py
Comment (till-m) [inline tests/test_target_space.py:190]: nitpick: One newline too many
```suggestion

```

File: bayes_opt/target_space.py
Comment (till-m) [inline bayes_opt/target_space.py:132]: @perezed00 @bwheelz36 @leandrobbraga @fmfn 
What do you think, does it make sense trigger warning here if we find values outside pbounds?

**Review by till-m — APPROVED** [review]
Overall I think this is basically ready for merge, excluding the 3 minor comments + the warning discussion.
@osullivryan If you have the time, I think a review from your side would also be appreciated.

**Comment by till-m** [comment]
Thanks for adding so much high-quality documentation, by the way!

File: bayes_opt/target_space.py
Comment (till-m) [reply bayes_opt/target_space.py]: Whoops, you already removed this. My bad!

**Comment by osullivryan** [comment]
Hey @perezed00  and @till-m , 

I took a look. Nice catch and fix! Awesome!

File: bayes_opt/target_space.py
Comment (perezed00) [reply bayes_opt/target_space.py:132]: There are three ways for parameters to exist beyond the pbounds:
1. Existing pbounds are shrunk using `target_space.set_bounds(small_pbounds)`
2. They are manually (i.e., intentionally) added using `target_space.register(x, y)`
3. Logs are loaded that don't respect the pbounds. 

I don't think that #1 should produce a warning since this intended behavior. To tackle #2 and #3 without affecting #1, we could issue a warning when a point outside the bounds is **registered**, rather than found?

File: tests/test_seq_domain_red.py
Comment (perezed00) [reply tests/test_seq_domain_red.py:190]: I found them helpful when writing the tests, since I could just "run" the test files, but I understand if you decide to remove them.

File: bayes_opt/target_space.py
Comment (till-m) [reply bayes_opt/target_space.py:132]: That sounds good, would you mind implementing the warning + adding a test for it?

File: tests/test_seq_domain_red.py
Comment (till-m) [reply tests/test_seq_domain_red.py:190]: let's keep them for now.

File: bayes_opt/target_space.py
Comment (perezed00) [reply bayes_opt/target_space.py:133]: This warning has been added and all warnings have been slightly updated for improved formatting.

**Comment by till-m** [comment]
From my side, I think this is good to go. @bwheelz36 @leandrobbraga @fmfn feel free to review. If you have no comments, I will merge this PR soon.

@perezed00 thank you very much for your contribution. As you may have noticed, we are a bit short on maintainers for this project currently. Based on the way you wrote this PR, the thoughtfulness of your approaches and the clarity of the code you wrote, I think it would be very valuable to have you on the team. If you're interested, please let me know :)

**Comment by perezed00** [comment]
@till-m Thanks for the invite to join the team. I accept, I would be happy to join. Just a bit of a warning however: I defend my dissertation in April, so I might be less responsive a couple months before then.

Cheers!

**Comment by bwheelz36** [comment]
Looks good to me. 
regarding the codecov checks, I'm not sure if you knew but you can check the code cov analysis of this PR here:
https://app.codecov.io/gh/bayesian-optimization/BayesianOptimization/pull/441
Apparently there are a few lines like `if (pbounds[0] > global_bounds[i, 1]):` which aren't being tested - should the test case be covering these?

**Pull request merged**
